Realm/Realm/Linq/RealmResultsVisitor2.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using Newtonsoft.Json;
+
+namespace Realms
+{
+    internal class RealmResultsVisitor2 : ExpressionVisitor
+    {
+        private readonly Realm _realm;
+        private readonly RealmObjectBase.Metadata _metadata;
+        private QueryModel _query;
+
+        // TODO Remove these fields when the CreateResults method is rewritten
+        private QueryHandle _coreQueryHandle;
+        private SortDescriptorHandle _sortDescriptor;
+
+        IQueryableCollection results;
+
+        internal RealmResultsVisitor2(Realm realm, RealmObjectBase.Metadata metadata)
+        {
+            _realm = realm;
+            _metadata = metadata;
+            _query = new QueryModel();
+            _query.WhereClauses = new List<WhereClause>();
+            _query.OrderingClauses = new List<OrderingClause>();
+        }
+
+        protected override Expression VisitMethodCall(MethodCallExpression node)
+        {
+            if (node.Method.DeclaringType == typeof(Queryable))
+            {
+                if (node.Method.Name == nameof(Queryable.Where))
+                {
+                    Visit(node.Arguments[0]);
+                    var whereClause = new WhereClauseVisitor(_metadata);
+                    var lambda = (LambdaExpression)StripQuotes(node.Arguments[1]);
+                    _query.WhereClauses.Add(whereClause.VisitWhere(lambda));
+                    return node;
+                }
+
+                if (IsSortClause(node.Method.Name))
+                {
+                    Visit(node.Arguments[0]);
+                    // Fix SOrtClause naming
```

Comment:
For all these things that are "TODO" more than comments, I would suggest to write //TODO. This way they're more visible and also can be found automatically in Visual Studio